Snippets/Core/Core_6/Handlers/DataAccess.cs

```diff
@@ -0,0 +1,156 @@
+﻿namespace Core5.Handlers
+{
+    using System;
+    using System.Threading.Tasks;
+    using NServiceBus;
+    using NServiceBus.Persistence;
+
+    public class DataAccess
+    {
+        public class Simple
+        {
+
+            #region BusinessData-Native-NotIdempotent
+
+            public class NonIdempotentHandler : IHandleMessages<AddOrderLine>
+            {
+                public IMyOrm Orm { get; set; }
+
+                public async Task Handle(AddOrderLine message, IMessageHandlerContext context)
+                {
+                    using (var session = Orm.OpenSession())
+                    {
+                        var order = session.Get(message.OrderId);
+                        order.AddLine(message.Product, message.Quantity);
+                        session.Commit();
+                    }
+                }
+            }
+
+            #endregion
+
+            #region BusinessData-Native-Idempotent
+
+            public class IdempotentHandler : IHandleMessages<AddOrderLine>
+            {
+                public IMyOrm Orm { get; set; }
+
+                public async Task Handle(AddOrderLine message, IMessageHandlerContext context)
+                {
+                    using (var session = Orm.OpenSession())
+                    {
+                        var order = session.Get(message.OrderId);
+                        if (order.HasLine(message.LineId))
+                        {
+                            return;
+                        }
+
+                        order.AddLine(message.Product, message.Quantity);
+                        session.Commit();
+                    }
+                }
+            }
+
+            #endregion
+        }
+
+        public class Managed
+        {
+            #region BusinessData-Native-Managed
+
+            public class IdempotencyEnforcer : IHandleMessages<OrderMessage>
+            {
+                public Task Handle(OrderMessage message, IMessageHandlerContext context)
+                {
+                    var session = context.SynchronizedStorageSession.MyOrmSession();
+                    var order = session.Get(message.OrderId);
+                    if (order.HasProcessed(context.MessageId))
```

Comment:
Well, not sure how to name it. The idea here is that when you want to implement the idempotency yourself, you need to somehow persist information about which message has been processed inside the entity itself.